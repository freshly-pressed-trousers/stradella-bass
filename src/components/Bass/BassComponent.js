import React from "react";
import styles from "./BassComponent.module.css";

import Button from "components/Button";
import ButtonRow from "components/ButtonRow";

const transpose = amount => note => note + amount;
const transformRootToMajorTriad = note => [note, note + 4, note + 7];
const transformRootToMinorTriad = note => [note, note + 3, note + 7];
const transformRootTo7th = note => [note + 10, note + 4, note + 7]; // 7th on the stradella bass ommits the root
const transformRootToDiminshed = note => [note + 3, note + 6, note + 9]; // Diminished on the stradella bass also ommits the root

const BASS_ROW_NOTES = [
	51, // Eb3
	46, // Bb2
	53, // F3
	48, // C3
	55, // G3
	50, // D3
	57, // A3
	52, // E3
];

// The counter bass in a stradella system is the major third above the root
const COUNTER_BASS_ROW_NOTES = BASS_ROW_NOTES.map(transpose(4));
// All the chord rows are one octave up from the bass row
const MAJOR_CHORD_ROW_NOTES = BASS_ROW_NOTES.map(transpose(12)).map(
	transformRootToMajorTriad
);
const MINOR_CHORD_ROW_NOTES = BASS_ROW_NOTES.map(transpose(12)).map(
	transformRootToMinorTriad
);
const SEVENTH_CHORD_ROW_NOTES = BASS_ROW_NOTES.map(transpose(12)).map(
	transformRootTo7th
);
const DIMINSHED_CHORD_ROW_NOTES = BASS_ROW_NOTES.map(transpose(12)).map(
	transformRootToDiminshed
);

export default function BassComponent() {
	return (
		<div className={styles.wrapper}>
			<ButtonRow label="Counter Bass">
				{COUNTER_BASS_ROW_NOTES.map(midiNote => (
					<Button key={midiNote} midiNote={midiNote} />
				))}
			</ButtonRow>
			<ButtonRow label="Bass">
				{BASS_ROW_NOTES.map(midiNote => (
					<Button key={midiNote} midiNote={midiNote} />
				))}
			</ButtonRow>
			<ButtonRow label="Major">
				{MAJOR_CHORD_ROW_NOTES.map((midiNotes, index) => (
					<Button key={index} midiNotes={midiNotes} />
				))}
			</ButtonRow>
			<ButtonRow label="Minor">
				{MINOR_CHORD_ROW_NOTES.map((midiNotes, index) => (
					<Button key={index} midiNotes={midiNotes} />
				))}
			</ButtonRow>
			<ButtonRow label="Seventh">
				{SEVENTH_CHORD_ROW_NOTES.map((midiNotes, index) => (
					<Button key={index} midiNotes={midiNotes} />
				))}
			</ButtonRow>
			<ButtonRow label="Diminished">
				{DIMINSHED_CHORD_ROW_NOTES.map((midiNotes, index) => (
					<Button key={index} midiNotes={midiNotes} />
				))}
			</ButtonRow>
		</div>
	);
}
