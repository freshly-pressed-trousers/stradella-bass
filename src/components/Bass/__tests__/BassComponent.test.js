import React from "react";
import BassComponent from "../BassComponent";
import { shallow } from "enzyme";

describe("BassComponent component", () => {
	it("matches snapshot", () => {
		const component = shallow(<BassComponent />);

		expect(component).toMatchSnapshot();
	});
});
