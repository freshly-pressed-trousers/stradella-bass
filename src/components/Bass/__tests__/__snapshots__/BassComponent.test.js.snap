// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BassComponent component matches snapshot 1`] = `
<div
  className="wrapper"
>
  <ButtonRowComponent
    label="Counter Bass"
  >
    <Connect(ButtonComponent)
      key="55"
      midiNote={55}
    />
    <Connect(ButtonComponent)
      key="50"
      midiNote={50}
    />
    <Connect(ButtonComponent)
      key="57"
      midiNote={57}
    />
    <Connect(ButtonComponent)
      key="52"
      midiNote={52}
    />
    <Connect(ButtonComponent)
      key="59"
      midiNote={59}
    />
    <Connect(ButtonComponent)
      key="54"
      midiNote={54}
    />
    <Connect(ButtonComponent)
      key="61"
      midiNote={61}
    />
    <Connect(ButtonComponent)
      key="56"
      midiNote={56}
    />
  </ButtonRowComponent>
  <ButtonRowComponent
    label="Bass"
  >
    <Connect(ButtonComponent)
      key="51"
      midiNote={51}
    />
    <Connect(ButtonComponent)
      key="46"
      midiNote={46}
    />
    <Connect(ButtonComponent)
      key="53"
      midiNote={53}
    />
    <Connect(ButtonComponent)
      key="48"
      midiNote={48}
    />
    <Connect(ButtonComponent)
      key="55"
      midiNote={55}
    />
    <Connect(ButtonComponent)
      key="50"
      midiNote={50}
    />
    <Connect(ButtonComponent)
      key="57"
      midiNote={57}
    />
    <Connect(ButtonComponent)
      key="52"
      midiNote={52}
    />
  </ButtonRowComponent>
  <ButtonRowComponent
    label="Major"
  >
    <Connect(ButtonComponent)
      key="0"
      midiNotes={
        Array [
          63,
          67,
          70,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="1"
      midiNotes={
        Array [
          58,
          62,
          65,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="2"
      midiNotes={
        Array [
          65,
          69,
          72,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="3"
      midiNotes={
        Array [
          60,
          64,
          67,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="4"
      midiNotes={
        Array [
          67,
          71,
          74,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="5"
      midiNotes={
        Array [
          62,
          66,
          69,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="6"
      midiNotes={
        Array [
          69,
          73,
          76,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="7"
      midiNotes={
        Array [
          64,
          68,
          71,
        ]
      }
    />
  </ButtonRowComponent>
  <ButtonRowComponent
    label="Minor"
  >
    <Connect(ButtonComponent)
      key="0"
      midiNotes={
        Array [
          63,
          66,
          70,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="1"
      midiNotes={
        Array [
          58,
          61,
          65,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="2"
      midiNotes={
        Array [
          65,
          68,
          72,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="3"
      midiNotes={
        Array [
          60,
          63,
          67,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="4"
      midiNotes={
        Array [
          67,
          70,
          74,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="5"
      midiNotes={
        Array [
          62,
          65,
          69,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="6"
      midiNotes={
        Array [
          69,
          72,
          76,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="7"
      midiNotes={
        Array [
          64,
          67,
          71,
        ]
      }
    />
  </ButtonRowComponent>
  <ButtonRowComponent
    label="Seventh"
  >
    <Connect(ButtonComponent)
      key="0"
      midiNotes={
        Array [
          73,
          67,
          70,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="1"
      midiNotes={
        Array [
          68,
          62,
          65,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="2"
      midiNotes={
        Array [
          75,
          69,
          72,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="3"
      midiNotes={
        Array [
          70,
          64,
          67,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="4"
      midiNotes={
        Array [
          77,
          71,
          74,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="5"
      midiNotes={
        Array [
          72,
          66,
          69,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="6"
      midiNotes={
        Array [
          79,
          73,
          76,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="7"
      midiNotes={
        Array [
          74,
          68,
          71,
        ]
      }
    />
  </ButtonRowComponent>
  <ButtonRowComponent
    label="Diminished"
  >
    <Connect(ButtonComponent)
      key="0"
      midiNotes={
        Array [
          66,
          69,
          72,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="1"
      midiNotes={
        Array [
          61,
          64,
          67,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="2"
      midiNotes={
        Array [
          68,
          71,
          74,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="3"
      midiNotes={
        Array [
          63,
          66,
          69,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="4"
      midiNotes={
        Array [
          70,
          73,
          76,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="5"
      midiNotes={
        Array [
          65,
          68,
          71,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="6"
      midiNotes={
        Array [
          72,
          75,
          78,
        ]
      }
    />
    <Connect(ButtonComponent)
      key="7"
      midiNotes={
        Array [
          67,
          70,
          73,
        ]
      }
    />
  </ButtonRowComponent>
</div>
`;
